hack/boilerplate/boilerplate.py

```diff
@@ -16,22 +16,46 @@
 
 from __future__ import print_function
 
+import argparse
 import json
 import mmap
 import os
 import re
 import sys
 
-def PrintError(*err):
-  print(*err, file=sys.stderr)
+parser = argparse.ArgumentParser()
+parser.add_argument("filenames", help="list of files to check, all files if unspecified", nargs='*')
+args = parser.parse_args()
 
-def file_passes(filename, extension, ref, regexs):
+rootdir = os.path.dirname(__file__) + "/../../"
+rootdir = os.path.abspath(rootdir)
+
+def get_ref(extension, refs):
+    if extension in refs:
+        return refs[extension]
+
+    ref_filename = rootdir + "hack/boilerplate/boilerplate." + extension + ".txt"
     try:
-        f = open(filename, 'r')
+        ref_file = open(ref_filename, 'r')
+        ref = ref_file.read().splitlines()
+        refs[extension] = ref
     except:
-        return False
+        refs[extension] = []
+
+    return refs[extension]
 
-    data = f.read()
+def file_passes(filename, refs, regexs):
+    extension = os.path.splitext(filename)[-1].strip().lower().split(".")[-1]
+    if extension == "":
+        return True
+
+    ref = get_ref(extension, refs)
```

Comment:
So you're going to read the boilerplate once per file? Wouldn't it make more sense to load all these up front, get the extensions we support, and filter out all the ones we don't support in `get_files`?